### PR TITLE
Apply compatible target conversion logic also to restore

### PR DESCRIPTION
**Pull request by pakrym**
Fix build failing during publish, because we restored with one set of frameworks and built with another.

Targets #2954

/cc @anurse @davidfowl @JunTaoLuo @troydai @moozzyk

### CODE REVIEW

File: src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs [hunk 1]
```diff
@@ -119,10 +119,31 @@ private bool EmitNupkg(PublishRoot root)
                 root.Operations.Delete(TargetPath);
             }
 
+            var selectedFrameworks = new List<FrameworkName>();
+
+            if (root.Frameworks.Any())
```
Comment (moozzyk) [inline src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs]: Is this check even needed?

File: src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs [hunk 2]
```diff
@@ -119,10 +119,31 @@ private bool EmitNupkg(PublishRoot root)
...
+                    selectedFrameworks.Add(selectedFramework.FrameworkName);
+                }
+            }
+
+
             // If this is a wrapper project, we need to generate a lock file before building it
             if (IsWrappingAssembly())
             {
-                var success = Restore(root, publishProject: this, restoreDirectory: project.ProjectDirectory)
+                var success = Restore(root, this, project.ProjectDirectory, selectedFrameworks)
```
Comment (moozzyk) [inline src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs]: the named parameters (especially `publishProject`) were useful here...

File: src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs [hunk 3]
```diff
@@ -140,19 +161,9 @@ private bool EmitNupkg(PublishRoot root)
 
             if (root.Frameworks.Any())
             {
-                // Make sure we only emit the nupkgs for the specified frameworks
-                // We need to pick actual project frameworks relevant to the publish
-                // but the project may not target exactly the right framework so we need
-                // to use a compatibility check.
-                foreach (var framework in root.Frameworks)
+                foreach (var selectedFramework in selectedFrameworks)
```
Comment (moozzyk) [inline src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs]: Again - `root.Frameworks.Any()` needed?

**Comment by moozzyk** [comment]
Is necessary to add a test for this or this is covered by existing tests?
Also - can you check why travis is failing and fix it?

**Comment by moozzyk** [comment]
Is necessary to add a test for this or this is covered by existing tests?
Also - can you check why travis is failing and fix it?

File: src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs [hunk 4]
```diff
@@ -119,10 +119,31 @@ private bool EmitNupkg(PublishRoot root)
...
+                    if (selectedFramework == null)
+                    {
+                        root.Reports.WriteError($"Unable to build {project.Name}. It is not compatible with the requested target framework: {framework.Key}");
+                        return false;
+                    }
+                    selectedFrameworks.Add(selectedFramework.FrameworkName);
+                }
+            }
+
+
```
Comment (troydai) [inline src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs]: Remove this empty line.

File: src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs [hunk 5]
```diff
@@ -437,7 +448,7 @@ private static string CombinePath(string path1, string path2)
             return Path.Combine(path1, path2);
         }
 
-        private async Task<bool> Restore(PublishRoot root, PublishProject publishProject, string restoreDirectory)
+        private async Task<bool> Restore(PublishRoot root, PublishProject publishProject, string restoreDirectory, IEnumerable<FrameworkName> frameworks)
```
Comment (troydai) [inline src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs]: Suggestion: I think the problem here is we have separate restore logic in publish which is different from `dnu restore`. Should we consider consolidate them together.

This is not supposed to block this PR.

Comment (troydai) [inline src/Microsoft.Dnx.Tooling/Publish/PublishProject.cs]: Name the new parameter `targetFrameworks`.

**Comment by davidfowl** [comment]
No tests?

**Comment by pakrym** [comment]
ping @anurse @davidfowl @JunTaoLuo @troydai @moozzyk  @cesarbs

File: test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPublishWrappedTests.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            DoPublish(sdk);
+        }
+
+        [ConditionalTheory, TraceTest]
+        [OSSkipCondition(OperatingSystems.Linux)]
+        [OSSkipCondition(OperatingSystems.MacOSX)]
+        [MemberData(nameof(ClrDnxSdks))] // project only supports dnx451
+        public void PublishWrappedProjectForSpecificRuntime(DnxSdk sdk)
+        {
+            DoPublish(sdk, "--runtime "+sdk.FullName);
```
Comment (JunTaoLuo) [inline test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPublishWrappedTests.cs]: Use string interpolation

File: test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPublishWrappedTests.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            var clrSdk = GetRuntime("clr", "win", "x86");
+
+            // Run it
+            var outputPath = Path.Combine(solution.ArtifactsPath, "approot", "src", "DnxConsoleApp");
+            var result = clrSdk.Dnx.Execute(
+                commandLine: $"--project \"{outputPath}\" --configuration Debug DnxConsoleApp run")
+                .EnsureSuccess();
+            Assert.Contains($"Hello from the wrapped project{Environment.NewLine}", result.StandardOutput);
+
+            TestUtils.CleanUpTestDir<DnuPublishTests>(sdk);
```
Comment (JunTaoLuo) [inline test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPublishWrappedTests.cs]: The class passed to `CleanUpTestDir` must match with the test class.

File: test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPublishWrappedTests.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        [MemberData(nameof(ClrDnxSdks))] // project only supports dnx451
+        public void PublishWrappedProjectForSpecificRuntime(DnxSdk sdk)
+        {
+            DoPublish(sdk, "--runtime "+sdk.FullName);
+        }
+
+        private static void DoPublish(DnxSdk sdk, string arguments = null)
+        {
+            var solutionName = Path.Combine("DnuWrapTestSolutions", "WrapAndPublish");
+            var solution = TestUtils.GetSolution<DnuPublishTests>(sdk, solutionName);
```
Comment (JunTaoLuo) [inline test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPublishWrappedTests.cs]: Generic must match test class.

**Comment by davidfowl** [comment]
:shipit:

**Comment by moozzyk** [comment]
@pakrym - looks like this change broke the build.

**Pull request merged**